### PR TITLE
[#42] Add support for dynamic supervisors

### PR DESCRIPTION
**Pull request by roman**
Closes (#42)

### Problem

As a developer, I want to spawn supervised worker routines dynamically.

### Solution

This commit offers the DynSupervisor API, which allows us to create a child-less Supervisor that can monitor any worker routine that is spawned in a procedural fashion.

### Changes

* Add DynSupervisor type
* Add Supervisor.GetCrashError
* Add startChildNode function to simplify the startChildNodes function
* Add ObserveDynSupervisor to test dynamic supervisors

### CODE REVIEW

File: cap/supervisor_options.go [hunk 1]
```diff
@@ -52,14 +52,19 @@ func WithNotifier(en EventNotifier) Opt {
...
 // records passed to this function are going to be supervised by the Supervisor
 // created from a SupervisorSpec.
 //
 // Check the documentation of NewSupervisorSpec for more details and examples.
 func WithNodes(nodes ...Node) BuildNodesFn {
-	emptyCleanupResources := func() error { return nil }
-	return func() ([]Node, CleanupResourcesFn, error) {
-		return nodes, emptyCleanupResources, nil
-	}
+	return withNodes(nodes)
```
Comment (julio-garcia-fc) [inline cap/supervisor_options.go:69]: I do not see the reason for this added indirection 🤷‍♂️

File: internal/stest/assertions.go [hunk 2]
```diff
@@ -124,10 +124,93 @@ func AssertPartialMatch(t *testing.T, evs []cap.Event, preds []EventP) {
...
+	// embedded in the ObserveDynSupervisor call
+	sup, err := cap.NewDynSupervisor(ctx, rootName, opts...)
+
+	if err != nil {
+		return evManager.Snapshot(), []error{err}
+	}
+
+	errors := []error{}
+
+	// start procedurally the children given children
```
Comment (julio-garcia-fc) [inline internal/stest/assertions.go:160]: s/given children//

File: internal/stest/assertions.go [hunk 3]
```diff
@@ -124,10 +124,93 @@ func AssertPartialMatch(t *testing.T, evs []cap.Event, preds []EventP) {
...
+	}
+
+	evIt := evManager.Iterator()
+
+	if len(errors) != 0 {
+		// once tests are done, we stop the supervisor
+		if err = sup.Terminate(); err != nil {
+			errors = append(errors, err)
+		}
+		evIt.SkipTill(SupervisorTerminated(rootName))
```
Comment (julio-garcia-fc) [inline internal/stest/assertions.go:175]: Would `WaitTill()` be a clearer name?

File: internal/stest/assertions.go [hunk 4]
```diff
@@ -124,10 +124,93 @@ func AssertPartialMatch(t *testing.T, evs []cap.Event, preds []EventP) {
...
+	}
+
+	// NOTE: We execute SkipTill to make sure all the supervision tree got started
+	// (or failed) before doing assertions/returning an error. Also, note we use
+	// ProcessName instead of ProcessStarted/ProcessFailed given that ProcessName
+	// matches an event in both success and error cases. The event from root must
+	// be the last event reported
+	evIt.SkipTill(ProcessName(rootName))
+
+	if err != nil {
```
Comment (julio-garcia-fc) [inline internal/stest/assertions.go:186]: I am lost. Where did this `err` come from? As best as I can tell. last time you set it and checked it, you returned?

File: internal/stest/assertions.go [hunk 5]
```diff
@@ -124,10 +124,93 @@ func AssertPartialMatch(t *testing.T, evs []cap.Event, preds []EventP) {
...
+		callback(sup, evManager)
+		return evManager.Snapshot(), []error{err}
+	}
+
+	// callback to do assertions with the event manager
+	callback(sup, evManager)
+
+	// once tests are done, we stop the supervisor
+	err = sup.Terminate()
```
Comment (julio-garcia-fc) [inline internal/stest/assertions.go:195]: You might consider using multiple context-named error variables. It will be easier to track.

**Review by julio-garcia-fc — APPROVED** [review]
LGTM'ish. I really like the dynamic supervisor.

File: cap/dyn_supervisor_test.go [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+		[]EventP{
+			SupervisorStarted("root"),
+			WorkerStarted("root/one"),
+			WorkerTerminated("root/one"),
+			SupervisorTerminated("root"),
+		})
+}
+
+// Test a supervision tree with three children start and stop in the default
+// order (LeftToRight)
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/dyn_supervisor_test.go:40]: The stop mechanism is in reverse order, so if your start is `LeftToRight`, your stop is going to be `RightToLeft`

File: cap/supervisor.go [hunk 7]
```diff
@@ -46,3 +54,113 @@ func (sup Supervisor) Wait() error {
...
+	supRuntimeName string,
+	mux *sync.Mutex,
+	terminated *bool,
+	terminateErr *error,
+	err error,
+	stopingTime time.Time,
+) {
+	mux.Lock()
+	defer mux.Unlock()
+	*terminated = true
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/supervisor.go:87]: Correct

File: cap/supervisor_options.go (hunk 1 above)
Comment (roman) [reply cap/supervisor_options.go:69]: Let me see if I can remove it, I was a bit tired when I did this and I may have a better way to go about it

File: cap/supervisor.go [hunk 8]
```diff
@@ -46,3 +54,113 @@ func (sup Supervisor) Wait() error {
...
+	*terminateErr = err
+	if err != nil {
+		eventNotifier.supervisorFailed(supRuntimeName, err)
+		return
+	}
+
+	// stopingTime is only relevant when we call the internal wait function
+	// from the Terminate() public API; if we just called from Wait(), we don't
+	// need to keep track of the stop duration
+	if stopingTime == (time.Time{}) {
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/supervisor.go:97]: Yep, golang and its quirks

File: internal/stest/assertions.go (hunk 3 above)
Comment (roman) [reply internal/stest/assertions.go:175]: I'll do this in a different PR, otherwise the changed LOC goes to the roof

File: cap/dyn_supervisor_test.go [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+
+	cancelWorker, err := sup.Spawn(WaitDoneWorker("one"))
+	assert.NoError(t, err)
+
+	err = sup.Terminate()
+	assert.NoError(t, err)
+
+	err = cancelWorker()
+	assert.Error(t, err)
+	assert.Equal(t, "could not talk to supervisor: send on closed channel", err.Error())
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/dyn_supervisor_test.go:390]: Added a new test on `supervisor_test.go` that asserts correct behavior on double termination

File: cap/dyn_supervisor_test.go [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+	events, errs := ObserveDynSupervisor(
+		context.TODO(),
+		"root",
+		[]cap.Node{
+			WaitDoneWorker("child0"),
+			WaitDoneWorker("child1"),
+			WaitDoneWorker("child2"),
+		},
+		[]cap.Opt{
+			cap.WithOrder(cap.RightToLeft),
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/dyn_supervisor_test.go:89]: I'll do this change on a different PR

File: cap/supervisor.go [hunk 11]
```diff
@@ -10,19 +10,60 @@ import (
...
+}
+
+// getTerminateErr is a (concurrent-safe) function that allows us to query if a
+// Supervisor was terminated.
+func (tm *terminationManager) getTerminateErr() (bool, error) {
+	tm.mux.Lock()
+	defer tm.mux.Unlock()
+	if tm.terminated {
+		// we already did the termination, just return the result
+		return tm.terminated, tm.terminateErr
```
Comment (julio-garcia-fc) [inline cap/supervisor.go:44]: This should be the only line in this function, I think? Unless you want to check for the broken `!tm.termimated && tm.terminatedErr != nil`

File: cap/dyn_supervisor.go [hunk 12]
```diff
@@ -9,36 +9,141 @@ import (
 	"github.com/capatazlib/go-capataz/internal/c"
 )
 
+// ctrlMsg executes some control logic on the Supervisor goroutine
```
Comment (julio-garcia-fc) [inline cap/dyn_supervisor.go:12]: I am glad you did this. I think it is easier to follow. I thought you might do this in a different PR.

**Review by julio-garcia-fc — COMMENTED** [review]
Very minor comments. Definitely LGTM

**Review by julio-garcia-fc — APPROVED** [review]
Pushed the wrong button, sorry

File: cap/dyn_supervisor_test.go [hunk 13]
```diff
@@ -90,12 +97,17 @@ func TestDynStartMutlipleChildrenRightToLeft(t *testing.T) {
...
+				// fashion
 				WorkerStarted("root/child0"),
 				WorkerStarted("root/child1"),
 				WorkerStarted("root/child2"),
+				// ^^^ start ordering is bound by children order on the
+				// ObserveDynSupervisor call
 				WorkerTerminated("root/child0"),
 				WorkerTerminated("root/child1"),
 				WorkerTerminated("root/child2"),
+				// ^^^ stop is left to right (first-in, first-out style)
```
[the remark this answers is not in the document]
Comment (roman) [reply cap/dyn_supervisor_test.go:110]: Thanks for the recommendation. I'm going to pass on it though. I'm following the OTP naming conventions here, and is a matter of getting used to the names and their behavior.

File: cap/supervisor.go (hunk 11 above)
Comment (roman) [reply cap/supervisor.go:44]: Yes, I think the `terminated` variable must be `true` for `terminateErr` to be filled... I think this is enough
